packages/app/auth/providers/earthdata-login.ts

```diff
@@ -0,0 +1,115 @@
+import type { OAuthConfig, OAuthUserConfig } from 'next-auth/providers/oauth'
+import type { TokenSetParameters } from 'openid-client'
+
+export interface EarthdataUser {
+    uid: string
+    first_name: string
+    last_name: string
+    email_address: string
+    registered_date: string
+    country: string
+    study_area: string
+    allow_auth_app_emails: boolean
+    user_type: string
+    affiliation: string
+    agreed_to_meris_eula: boolean
+    agreed_to_sentinel_eula: boolean
+    email_verified: boolean
+    user_authorized_apps: number
+}
+
+export const basePath = 'https://urs.earthdata.nasa.gov'
+
+export type EDLTokenSetParameters = Pick<
+    TokenSetParameters,
+    'access_token' | 'token_type' | 'refresh_token'
+> & {
+    expires_in: number
+    endpoint: string
+}
+/**
+ * A custom OAuth2 Provider for authenticating to Earthdata Login
+ * (https://urs.earthdata.nasa.gov, https://next-auth.js.org/configuration/providers/oauth)
+ */
+export default function EarthdataLoginProvider<P extends EarthdataUser = any>(
+    options: OAuthUserConfig<P>
+): OAuthConfig<P> {
+    return {
+        id: 'earthdata-login',
+        name: 'Earthdata Login',
+        version: '2.0',
+        type: 'oauth',
+        authorization: {
```

Comment:
oAuth authorization URL, not much different here from normal oAuth implementations